Dias fails to run after egg_conductor due to folder name

CNV calling doesn’t work without single in directory name


`$ dias.py -a CEN cnvcall /output/CEN-230407_1423/ CEN_0173_230412_CNV_exclude.txt
Traceback (most recent call last):
  File "/mnt/storage/apps/software/dias/1.10.2/dias.py", line 197, in <module>
    main()
  File "/mnt/storage/apps/software/dias/1.10.2/dias.py", line 174, in main
    args.sample_list
  File "/mnt/storage/apps/software/dias/1.10.2/cnvcalling.py", line 41, in run_cnvcall_app
    "must contain 'single'".format(ss_workflow_out_dir)
AssertionError: Path '/output/CEN-230407_1423/' is not an accepted directory, must contain 'single'`

To solve I changed directory name to /output/CEN-230407_1423_single/, moving forward we could remove the check for 'single' from dias.py.

--- a/ReadMe.md
+++ b/ReadMe.md
@@ -1,10 +1,21 @@
 # dias.py
 
-This readme is mainly to help understand what's going on. To understand how to run it, get trained by a qualified member of the team or read the Dias manual in Q-Pulse
+This Python module was developed by the East GLH bioinformatics team to run DNAnexus workflows as part of germline analysis of next generation sequencing data: from FASTQ to Excel variant workbooks, generating quality and coverage reports.
+This READme describes the module architecture and should serve as a guide to future users and developers of the Dias pipeline.
+To understand how to run it, get trained by a qualified member of the team and read the Dias manual.
 
 ## Function
 
-Run the Dias pipeline stored in DNAnexus
+Run components of the Dias pipeline stored in DNAnexus as apps or workflows. This module collects the aproppriate input files for each stage in the workflows, sets off the jobs on DNAnexus and direct the output files into folders in a predefined structure.
+
+
+It curently supports the following workflows and apps:
+* dias_single_workflow: ran once per sample on a sequencing run
+* dias_multi_workflow: ran once per sequencing run
+* qc: the MultiQC app is ran once per sequencing run
+* cnvcall: GATK's germline CNV calling steps are ran in cohort mode (ran once per sequencing run)
+* dias_reports_workflow: ran once per sample
+* dias_cnvreports_workflow: ran once per sample
 
 ## Files
 
@@ -17,10 +28,10 @@ Run the Dias pipeline stored in DNAnexus
 - multi_workflow.py
   - Generates the batch tsv and runs the dx cmd to start the multi workflow
 - multiqc.py
-  - Runs the multiqc
+  - Runs the MultiQC app to generate a run-level quality report
 - cnvcalling.py
-  - Generate batch files for CNVcalling
+  - Runs the GATK_gCNV app for run-level CNVcalling, with samples listed in the input file excluded
 - reports.py
-  - Generate batch files for reports and reanalysis
+  - Generate batch files for reports workflow and reanalysis requests
 - cnvreports.py
-  - Generate batch files for CNV reports and CNV reanalysis
+  - Generate batch files for CNV reports workflow and CNV reanalysis requests

--- a/cnvcalling.py
+++ b/cnvcalling.py
@@ -32,14 +32,8 @@ def run_cnvcall_app(ss_workflow_out_dir, dry_run, assay_config, assay_id, exclud
     project_name = get_dx_cwd_project_name()
 
     # Make sure path provided is an actual ss workflow output folder
-    assert ss_workflow_out_dir.startswith("/"), (
-        "Input directory must be full path (starting at /)")
-    path_dirs = [x for x in ss_workflow_out_dir.split("/") if x]
-    is_path_single = [ele for ele in path_dirs if "single" in ele]
-    assert is_path_single != [], (
-        "Path '{}' is not an accepted directory, "
-        "must contain 'single'".format(ss_workflow_out_dir)
-    )
+    assert ss_workflow_out_dir.startswith("/output/"), (
+        "Input directory must be full path (starting with /output/)")
 
     # Find the app name and create an output folder for it under ss
     app_name = get_object_attribute_from_object_id_or_path(

--- a/cnvcalling.py
+++ b/cnvcalling.py
@@ -51,31 +51,23 @@ def run_cnvcall_app(ss_workflow_out_dir, dry_run, assay_config, assay_id, exclud
         bambi_files.extend(find_files(project_name, folder_path, ext))
 
     # Read in list of samples that did NOT PASS QC
-    if excluded_sample_list is None:
-        sample_names = []
-    else:
-        sample_names = []
-        # parse sample exclusion file
-        with open(excluded_sample_list) as fh:
-            for line in fh:  # line can be a sample name or sample tab panel name
-                sample_names.append(line.strip().split("\t")[0])
+    excluded_sample_names = []
+    # parse sample exclusion file
+    with open(excluded_sample_list) as fh:
+        for line in fh:  
+            # line can be a sample name or sample tab panel name
+            # remove file extensions after first "_"
+            excluded_sample_names.append(line.strip().split("\t")[0].split('_')[0])
 
-    # Check that the sample list is not just the first field but it
-    # is until EGG
-    last_field = re.compile("-EGG[0-9]")
-    for sample in sample_names:
-        match = re.search(last_field, sample)
-        if match is None:
-            raise Exception("sample '{}' is not full sample name "
-                            "up to EGG code".format(
-                                sample
-                                ))
-    # Keep the excluded sample name up to EGG and remove anything after 
-    # that which would be after "_". This is if the excluded list
-    # contains up full filenames up to .bam
-    sample_names = [x.split('_')[0] for x in sample_names]
-    # Remove bam/bai files of QC failed samples
-    sample_bambis = [x for x in bambi_files if x.split('_')[0] not in sample_names]
+    # Remove bam/bai files of excluded (QC failed) samples
+    sample_bambis = [x for x in bambi_files if x.split('_')[0] not in excluded_sample_names]
+    print(
+        "{} out of {} samples were excluded from CNV calling".format(
+            (len(bambi_files) - len(sample_bambis)) / 2, len(bambi_files) / 2)
+        )
+    if len(sample_bambis) < 60:
+        print("Less than 30 samples suitable for CNV calling, \n \
+        which is below the threshold for optimal performance")
 
     # Find the file-IDs of the passed bam/bai samples
     file_ids = ""

--- a/cnvcalling.py
+++ b/cnvcalling.py
@@ -1,9 +1,5 @@
 #!/usr/bin/python
-from ast import excepthandler
-import sys
 import subprocess
-import dxpy
-import re
 
 from general_functions import (
     get_dx_cwd_project_name,
@@ -39,8 +35,6 @@ def run_cnvcall_app(ss_workflow_out_dir, dry_run, assay_config, assay_id, exclud
     app_name = get_object_attribute_from_object_id_or_path(
         assay_config.cnvcall_app_id, "Name"
     )
-
-
     app_output_dir = make_app_output_dir(assay_config.cnvcall_app_id, ss_workflow_out_dir, app_name, assay_id)
 
     # Find bam and bai files from sentieon folder

--- a/cnvcalling.py
+++ b/cnvcalling.py
@@ -91,14 +91,20 @@ def run_cnvcall_app(ss_workflow_out_dir, dry_run, assay_config, assay_id, exclud
         file_ids,
         project_name, app_output_dir
     )
+
     # upload excluded regions file
     excluded_list_path = app_output_dir + "/" + "excluded_list.tsv"
     cmd = "dx upload {} --path {}".format(excluded_sample_list, excluded_list_path)
-    subprocess.check_output(cmd, shell=True)
 
     if dry_run is True:
+        print("Created output dir: {}".format(app_output_dir))
         print("Final cmd ran: {}".format(command))
+        print("Deleting '{}' as part of the dry-run".format(app_output_dir))
+        delete_folders_cmd = "dx rm -r {}".format(app_output_dir)
+        subprocess.call(delete_folders_cmd, shell=True)
+
     else:
         subprocess.call(command, shell=True)
+        subprocess.check_output(cmd, shell=True)
 
     return app_output_dir

--- a/dias.py
+++ b/dias.py
@@ -215,6 +215,7 @@ def main():
     # Set off relevant workflow based on subcommand
     # with applicable inputs
     if subcommand == "cnvcall":
+        assert args.sample_list, "Please specify a sample exclusion list, including at least the control sample"
         cnvcall_applet_out_dir = run_cnvcall_app(
             args.input_dir, args.dry_run, config, assay_id,
             args.sample_list

